File: 433-minimum-genetic-mutation/minimum-genetic-mutation.cs
public class Solution {
    public int MinMutation(string startGene, string endGene, string[] bank) {
        HashSet<string> geneBank = new HashSet<string>(bank);
        if (!geneBank.Contains(endGene)) {
            return -1; 
        }
        
        char[] geneChars = new char[] { 'A', 'C', 'G', 'T' };
                Queue<(string gene, int steps)> queue = new Queue<(string, int)>();
        HashSet<string> visited = new HashSet<string>(); 
        queue.Enqueue((startGene, 0));
        visited.Add(startGene);
        
        while (queue.Count > 0) {
            var (currentGene, steps) = queue.Dequeue();
                        for (int i = 0; i < currentGene.Length; i++) {
                foreach (char c in geneChars) {
                    if (c != currentGene[i]) { 
                        char[] mutatedGeneArr = currentGene.ToCharArray();
                        mutatedGeneArr[i] = c;
                        string mutatedGene = new string(mutatedGeneArr);
                        
                        if (mutatedGene == endGene) {
                            return steps + 1;
                        }
                        
                        if (geneBank.Contains(mutatedGene) && !visited.Contains(mutatedGene)) {
                            queue.Enqueue((mutatedGene, steps + 1));
                            visited.Add(mutatedGene); 
                        }
                    }
                }
            }
        }
                return -1;
    }
}

File: 433-minimum-genetic-mutation/minimum-genetic-mutation.js
/**
 * @param {string} startGene
 * @param {string} endGene
 * @param {string[]} bank
 * @return {number}
 */
var minMutation=function(startGene, endGene, bank){
    const isValidMutation=(gene1, gene2)=>{
        let count=0; 
        for (let i=0; i<gene1.length; i++){
            if(gene1[i]!==gene2[i]) count++;
            if (count >1) return false;
        }return count===1;
    };
    const queue=[[startGene,0]];
    const visited=new Set();
    while (queue.length >0){
        const [currentGene, mutations]=queue.shift();
        if(currentGene===endGene) return mutations;
        for(const gene of bank){
            if(!visited.has(gene) && isValidMutation(currentGene, gene)){
                visited.add(gene);
                queue.push([gene,mutations+1]);
            }            
        }
    }return -1;
}